common/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/RowReaderThread.java

```diff
@@ -53,23 +54,34 @@ public RowReaderThread(
 
     @Override
     protected void createActualReader() throws PulsarClientException, ExecutionException {
-        reader = CachedPulsarClient
+        final ReaderBuilder<?> readerBuilder = CachedPulsarClient
                 .getOrCreate(clientConf)
                 .newReader(schema)
-                .topic(topic)
+                .topic(topicRange.getTopic())
                 .startMessageId(startMessageId)
                 .startMessageIdInclusive()
-                .loadConf(readerConf)
-                .create();
+                .loadConf(readerConf);
+        if (!topicRange.isFullRange()) {
+            readerBuilder.keyHashRange(topicRange.getPulsarRange());
+        }
+        reader = readerBuilder.create();
     }
 
     @Override
     protected void emitRecord(Message<?> message) throws IOException {
-        MessageId messageId = message.getMessageId();
-        Row record = deserializer.deserialize(message);
-        if (deserializer.isEndOfStream(record)) {
-            return;
+        try {
+            MessageId messageId = message.getMessageId();
+            Row record = deserializer.deserialize(message);
+            if (deserializer.isEndOfStream(record)) {
+                return;
+            }
+            if (record.getField(0) == null){
```

Comment:
This change seems to break reading from topics with certain schemas (nullable first fields). Are there other changes I'm missing or are we in the process of deprecating the old stream-reading behavior?

Reply:
I can open an issue if you think that's a better place to discuss this, but I chose here as this code hasn't been officially released yet.

Reply:
Thank you for your support.
The code here needs to be deleted. Earlier, when testing the function, because the message was always lost, I added a check here. Later, the bug was fixed in pr #139.

The Sub-stream reader distribution range is wrong, and I am repairing it. This check will be removed.